### PR TITLE
Optional nuclide-based isotropic in lab scattering

### PR DESCRIPTION
**Pull request by wbinventor**
This pull request introduces the option to use isotropic-in-lab scattering on a nuclide-by-nuclide basis. The majority of the new code was graciously contributed by @walshjon at my request, with a few very minor additions of my own (to the Python API and documentation).

There is a new optional ``scattering`` attribute/sub-element for none, some or all nuclides in "materials.xml". The ``scattering`` attribute may be set to "ace" to use the scattering laws specified in the ACE files. Of course this is what is currently done, and will continue to be done by default. Alternatively, when set to "lab", the ACE scattering laws are used to sample the outgoing energy, but an isotropic-in-lab distribution is used to sample the outgoing angle at each scattering interaction (elastic or inelastic). 

I understand that there will likely be a lot of questions about why a user would ever want the capability to turn off this physics in the code. The reason we implemented this option is to do comparisons between Monte Carlo and deterministic codes. In particular, one may use OpenMC to compute multi-group cross-sections for a deterministic code. In this case, the ``scattering`` attribute may be used to quantify the effects of anisotropic scattering and the effect of applying various types of transport corrections to the total cross-section. Perhaps the other developers will decide that this feature is not useful enough to integrate into the core code - in this case, I still thought it might be helpful to submit this PR for anyone else out there who may find this option useful.

### CODE REVIEW

**Comment by paulromano** [comment]
Let me cogitate on the value of being able to turn off physics; I do appreciate the argument regarding comparison to deterministic codes. I could be convinced either way that this should or shouldn't go in to the develop branch.

An implementation issue -- `list_iso_lab` could very well be `type(ListInt)` with values of 0 and 1 instead of .false. and .true. That would circumvent having to implement yet another list.

**Comment by nelsonag** [comment]
Hey guys, two quick questions:
1. How strong of a need is there to make the isotropic feature material dependent?
  a. I ask this one because if the option was not material dependent it could be implemented in a much simpler fashion which would add less complexity and baggage to the rest of the code.  
2. When using this for your studies (I am familiar with Will's work using this feature at least), would it be useful to maintain the same energy transfer matrices?
  a. I ask this because if you sample mu once more after the normal scatter routines are run, then you will still preserve the energy transfer probabilities, but simply overwrite the old mu with the new isotropic in lab mu.  This seems like a desirable property (i.e., turning this feature on or off would not change the infinite homogenous media keff if written in this way), but also allows for using the feature with S(a,b) data.

**Comment by walshjon** [comment]
I agree that there are diminishing returns in making the option available nuclide-by-nuclide, material-by-material, rather than just doing it globally, nuclide-by-nuclide.  My thought was that this was mostly going to be a research hack of a feature and it couldn't really hurt to have that sort of flexibility.  Also, from a user input perspective, it almost seems cleaner - that is, relative to the alternative of having a `scattering` attribute associated with a `nuclide` element in a single material apply globally to all `nuclide` elements associated with the same the same nuclide across all materials.  From talking to @wbinventor, it does seem that having at least the global nuclide-by-nuclide specification may be helpful.

I agree that working it so that S(a,b) data is permissible with the lab isotropic scattering, if desired, would be a better, more flexible way to go.

All this said, I likely won't use this feature very much at all and will defer to the MG XS generation guys.

**Comment by wbinventor** [comment]
@paulromano - I'll go ahead and make the switch to ``ListInt`` to eliminate the need for a new type of list. Of course, I understand that need for this PR is up for debate, so this change will only be beneficial if we do indeed decide to merge this into develop at some point.

The ability to set isotropric in lab scattering on a nuclide-by-nuclide basis isn't entirely necessary, and perhaps it would be best to simply have a global setting to set it for all nuclides/materials. The original idea was to keep it fine-grained on a nuclide basis so that users can easily quantify transport effects from each nuclide in a problem. If this is the dealbreaker to incorporating the code into develop, then I can make the changes to make it a global setting instead.

**Comment by wbinventor** [comment]
I removed the ``ListLog`` type and replaced it with the standard ``ListInt`` type for the isotropic-in-lab flags per your suggestion @paulromano. @walshjon updated this PR such that it will preserve the energy transfer matrices with S(a,b) data per your request @nelsonag.

File: docs/source/usersguide/input.rst [hunk 1]
```diff
@@ -1116,6 +1116,15 @@ Each ``material`` element can have the following attributes or sub-elements:
     .. note:: If one nuclide is specified in atom percent, all others must also
               be given in atom percent. The same applies for weight percentages.
 
+    An optional attribute/sub-element for each nuclide is ``scattering``. This
+    attribute may be set to "ace" to use the scattering laws specified in the
+    ACE files (default). Alternatively, when set to "lab", the ACE scattering
```
Comment (paulromano) [inline docs/source/usersguide/input.rst]: I think "isotropic" might be a bit more obvious than "lab". I also don't really like "ace" because there will probably be a day when we no longer use ACE cross sections, but not sure what a better identifier is.

**Comment by paulromano** [comment]
Unless there is a good reason not to, it would be good to support setting isotropic scattering for elements in addition to nuclides.

File: src/physics.F90 [hunk 2]
```diff
@@ -459,11 +470,17 @@ subroutine elastic_scatter(i_nuclide, rxn, E, uvw, mu_lab, wgt)
...
 
     ! compute cosine of scattering angle in LAB frame by taking dot product of
-    ! neutron's pre- and post-collision angle
-    mu_lab = dot_product(uvw, v_n) / vel
+    ! neutron's pre- and post-collision unit vectors
+    if (iso_lab) then
+      uvw(1) = TWO * prn() - ONE
+      phi = TWO * PI * prn()
+      uvw(2) = cos(phi) * sqrt(ONE - uvw(1)*uvw(1))
+      uvw(3) = sin(phi) * sqrt(ONE - uvw(1)*uvw(1))
```
Comment (paulromano) [inline src/physics.F90]: It seems like the logic for isotropic scattering ought to be at the end of `subroutine scatter` rather than inside each of `elastic_scatter`, `sab_scatter`, and `inelastic_scatter`. If we did that, those three routines wouldn't need any modifications (I think?). Also, we can move the calculation of mu_lab to the end of `scatter` as well.

**Comment by wbinventor** [comment]
Thanks for the latest comments @paulromano, I'll try to address them later this week. Do you have a newfound interest in this capability? I still need this capability for my work, but didn't know if anyone else was interested in it at the moment.

Unfortunately, a bug crept in when I merged with the v0.7, and now a particle's energy may exceed 20 MeV once in a blue moon. I haven't fixed this yet since I've been busy with other things, but need to resolve it for my own research. If this is a feature you think would be interesting to include in the release version of the code (esp. in light of PR #471) then I'll jump on this sooner rather than later.

**Comment by paulromano** [comment]
@wbinventor See the discussion on the [dev list](https://groups.google.com/forum/#!topic/openmc-dev/oJlWqY8mMsc).

File: src/physics.F90 (hunk 2 above)
Comment (wbinventor) [inline src/physics.F90]: @walshjon what do you think about @paulromano's proposal here? Does this make sense or do you see any issues with it?

File: docs/source/usersguide/input.rst (hunk 1 above)
Comment (wbinventor) [inline docs/source/usersguide/input.rst]: I agree that "lab" is not a very intuitive descriptor. I also think that "isotropic" is not particularly intuitive since it may be with respect to either the lab or CM frames (and most users might assume CM). I just pushed a commit which uses "iso-in-lab" instead. What do you think of this?

Comment (wbinventor) [inline docs/source/usersguide/input.rst]: Also, if you have a particular suggestion of how you would rephrase this description without referencing ACE then I will use it; otherwise, I'm unsure of what would work well here.

File: src/physics.F90 (hunk 2 above)
Comment (wbinventor) [inline src/physics.F90]: I went ahead and consolidated the iso-in-lab scattering logic into `subroutine scatter` such that the `elastic_scatter`, `sab_scatter` and `inelastic_scatter` routines now remain the same as before. I would certainly like both @paulromano and @walshjon at the least to look at this to make sure that the new (greatly simplified) implementation will still perform as expected.

**Comment by wbinventor** [comment]
@paulromano I agree that this feature should be supported for not only nuclides but also elements. Since natural elements are "expanded" into nuclides, do you suggest setting the "iso-in-lab" tag for each of the individual nuclides that are initialized during the expansion?

**Comment by wbinventor** [comment]
@paulromano I have addressed all of your comments. The iso-in-lab feature is now supported for both nuclides as well as natural elements. The iso-in-lab scattering is consolidated directly into the single `scatter` subroutine in "physics.F90." Since I'm really just hacking away here with little prior knowledge of "physics.F90," I hope that both you and @walshjon can review the latest version to ensure that everything checks out.

File: src/physics.F90 [hunk 3]
```diff
@@ -390,6 +399,20 @@ subroutine scatter(p, i_nuclide)
...
+
+      ! sample isotropic-in-lab outgoing direction
+      uvw_new(1) = TWO * prn() - ONE
+      phi = TWO * PI * prn()
+      uvw_new(2) = cos(phi) * sqrt(ONE - uvw_new(1)*uvw_new(1))
+      uvw_new(3) = sin(phi) * sqrt(ONE - uvw_new(1)*uvw_new(1))
+      mu_lab = dot_product(uvw_old, uvw_new)
+
+      ! change direction of particle
+      p % coord(1) % uvw = rotate_angle(uvw_new, mu_lab)
```
Comment (walshjon) [inline src/physics.F90]: `uvw_new` already holds the correct values for the particle's `uvw`.  The rotate function would resample `phi` inconsistently with the calculated `mu_lab` and should take `uvw_old` as an argument.  The latter is not really germane; `p % coord(1) % uvw` can just be set to `uvw_new`.

Also, `p % mu` should be set to `mu_lab` here.  Otherwise, it still holds the 'true physics' value computed by whichever scattering subroutine was called above.

Comment (wbinventor) [inline src/physics.F90]: @walshjon - I *think* that my latest commit revises this code block as you've advised.

Comment (walshjon) [inline src/physics.F90]: @wbinventor , yes, this is what I had in mind.

**Comment by wbinventor** [comment]
Great, then I think this is ready for final review by @paulromano now.

File: docs/source/usersguide/input.rst [hunk 4]
```diff
@@ -1136,6 +1136,15 @@ Each ``material`` element can have the following attributes or sub-elements:
     .. note:: If one nuclide is specified in atom percent, all others must also
               be given in atom percent. The same applies for weight percentages.
 
+    An optional attribute/sub-element for each nuclide is ``scattering``. This
+    attribute may be set to "ace" to use the scattering laws specified in the
```
Comment (paulromano) [inline docs/source/usersguide/input.rst]: How about "data" or "library" instead of "ace"? That way we won't have to change it in our post-ACE future.

File: src/input_xml.F90 [hunk 5]
```diff
@@ -1934,6 +1937,21 @@ subroutine read_materials_xml()
           end if
         end if
 
+        ! Check enforced isotropic lab scattering
+        if (check_for_node(node_nuc, "scattering")) then
+          call get_node_value(node_nuc, "scattering", temp_str)
+          if (trim(adjustl(to_lower(temp_str))) == "iso-in-lab") then
```
Comment (paulromano) [inline src/input_xml.F90]: No need to call `trim` function

File: src/input_xml.F90 [hunk 6]
```diff
@@ -1934,6 +1937,21 @@ subroutine read_materials_xml()
...
+            call list_iso_lab % append(1)
+          else if (trim(adjustl(to_lower(temp_str))) == "ace") then
```
Comment (paulromano) [inline src/input_xml.F90]: No need to call `trim` function

File: src/input_xml.F90 [hunk 7]
```diff
@@ -2014,6 +2035,29 @@ subroutine read_materials_xml()
...
+        ! Check enforced isotropic lab scattering
+        if (check_for_node(node_ele, "scattering")) then
+          call get_node_value(node_ele, "scattering", temp_str)
+        else
+          temp_str = "ace"
+        end if
+
+        ! Set ace or iso-in-lab scattering for each nuclide in element
+        do k = 1, n_nuc_ele
+          if (trim(adjustl(to_lower(temp_str))) == "iso-in-lab") then
```
Comment (paulromano) [inline src/input_xml.F90]: No need to call `trim` function

File: src/input_xml.F90 [hunk 8]
```diff
@@ -2014,6 +2035,29 @@ subroutine read_materials_xml()
...
+            call list_iso_lab % append(1)
+          else if (trim(adjustl(to_lower(temp_str))) == "ace") then
```
Comment (paulromano) [inline src/input_xml.F90]: and again..

**Comment by paulromano** [comment]
@wbinventor Looks much better now with the iso-in-lab logic only in `scatter`. Last thing I can think of is that you need to update the RELAX NG schemata so that our validators still work.

File: docs/source/usersguide/input.rst (hunk 4 above)
Comment (wbinventor) [inline docs/source/usersguide/input.rst]: Sounds good, I've migrated to "data" instead of "ace" now.

**Comment by wbinventor** [comment]
@paulromano I've updated this per all of your comments. On the rare occasion that I have to update the RELAX NG schema I always find myself to searching around to find the command to build the new schema. This might be something for us to consider adding to our developer's guide. (even just the command `trang file.rnc file.rng`)
